src/decoder.rs

```diff
@@ -80,10 +83,16 @@ impl Decoder {
         }
     }
 
-    pub fn decode_raw<M: Memory>(&self, memory: &mut M, pc: u64) -> Result<Instruction, Error> {
+    pub fn decode_raw<M: Memory>(&mut self, memory: &mut M, pc: u64) -> Result<Instruction, Error> {
+        let instruction_cache_key = pc as usize % INSTRUCTION_CACHE_SIZE;
```

Comment:
Due to RISC-V instruction encoding, the lowest bit in PC will always be zero, so one trivial work we can do is:

`(pc >> 1) as usize % INSTRUCTION_CACHE_SIZE`

Otherwise half of the cache will always be empty.